homeassistant/components/zwave_js/__init__.py

```diff
@@ -293,15 +315,15 @@ async def async_on_value_added(
         )
 
     @callback
-    def async_on_node_removed(node: ZwaveNode) -> None:
+    def async_on_node_removed(event: dict) -> None:
         """Handle node removed event."""
+        node: ZwaveNode = event["node"]
+        replaced: bool = event.get("replaced", False)
         # grab device in device registry attached to this node
         dev_id = get_device_id(client, node)
         device = dev_reg.async_get_device({dev_id})
-        # note: removal of entity registry entry is handled by core
-        dev_reg.async_remove_device(device.id)  # type: ignore
-        registered_unique_ids.pop(device.id, None)  # type: ignore
-        discovered_value_ids.pop(device.id, None)  # type: ignore
+        if not replaced:
+            remove_device(device)  # type: ignore
```

Comment:
If we know that the device entry exists we should assert it before using it.

Reply:
agreed, I just assumed we wanted to keep the same pattern 👍🏾

Reply:
It's better to assert so we can remove the type ignore.